### PR TITLE
Feat ssrc rewriting

### PR DESCRIPTION
**Pull request by jallamsetty1**
In this feature:

- There will be a fixed set of remote tracks and therefore a fixed set of SSRCs.

- The bridge will re-map those SSRCs to a remote sender based on what is being requested and then the owner (source) of a remote track will change.

- The bridge does the remapping based on the receiver constraints received from the client.

- However, the receiver constraints currently are derived from the tracks in redux but with ssrc-rewriting, the tracks won't be created unless it is signaled. So it is a chicken and egg problem. Therefore, we will be constructing the receiver constraints based on the remote presence since all the source information is already signaled in presence.

There is jitsi-meet PR which does the receiver constraints part - https://github.com/jitsi/jitsi-meet/pull/12408/files#diff-ce724de6e7f229a65b2483d045183950192efdb68e35f9c5587a139609a49de6L371

### CODE REVIEW

File: JitsiConference.js [hunk 1]
```diff
@@ -156,7 +157,12 @@ export default function JitsiConference(options) {
     this.eventEmitter = new EventEmitter();
     this.options = options;
     this.eventManager = new JitsiConferenceEventManager(this);
-    this.participants = {};
+
+    /**
+     * List of all the participants in the conference.
+     * @type {Map<string, JitsiParticipant>};
+     */
+    this.participants = new Map();
```
Comment (hristoterezov) [inline JitsiConference.js:165]: Out of curiosity why did you change it to `Map`?

File: JitsiParticipant.js [hunk 2]
```diff
@@ -294,6 +308,32 @@ export default class JitsiParticipant {
...
 
+    /**
+     * Sets source info.
+     * @param {MediaType} mediaType The media type, 'audio' or 'video'.
+     * @param {boolean} muted The new muted state.
+     * @param {string} sourceName The name of the source.
+     * @param {string} videoType The video type of the source.
+     * @returns {void}
+     */
+    setSources(mediaType, muted, sourceName, videoType) {
```
Comment (hristoterezov) [inline JitsiParticipant.js:319]: Should we rename it to `_setSources` in order to indicate that this method is not part of the public API of LJM? WDYT?

File: modules/xmpp/SignalingLayerImpl.js [hunk 3]
```diff
@@ -180,12 +180,14 @@ export default class SignalingLayerImpl extends SignalingLayer {
                 = this._remoteSourceState[endpointId] || (this._remoteSourceState[endpointId] = {});
 
             for (const sourceName of Object.keys(sourceInfoJSON)) {
+                let presenceChanged = false;
```
Comment (hristoterezov) [inline modules/xmpp/SignalingLayerImpl.js:183]: Isn't `sourcesChanged` more suitable name?

**Comment by jallamsetty1** [comment]
> I'm wondering why do we decided to send this information trough presence and not trough the bridge channels?
> 
> Could you please give example of what do we expect to receive trough presence?
> 
> I was thinking that this feature will include:
> 
> * having a fixed set of sources per receiver.
> * reusing those sources for different participants.
> 
> Is this different feature? If it is different could you please add some details.

This information(mute state, video type) has always been sent through presence. All I am doing now is saving the source information (mute state, videoType and source name) as a member of JitsiParticipant whenever presence from remote participant is received.

In this feature:

- There will be a fixed set of remote tracks and therefore a fixed set of SSRCs.
- The bridge will re-map those SSRCs to a remote senders based on what is being requested and then the owner (source) of a remote track will change.
- The bridge does the remapping based on the receiver constraints received from the client. 
- However, the receiver constraints currently are derived from the tracks in redux but with ssrc-rewriting, the tracks won't be created unless it is signaled. So it is a chicken and egg problem. Therefore, we will be constructing the receiver constraints based on the remote presence since all the source information is already signaled in presence.
There is jitsi-meet PR which does this - https://github.com/jitsi/jitsi-meet/pull/12408/files#diff-ce724de6e7f229a65b2483d045183950192efdb68e35f9c5587a139609a49de6L371

File: JitsiParticipant.js (hunk 2 above)
Comment (jallamsetty1) [reply JitsiParticipant.js]: Done

File: modules/xmpp/SignalingLayerImpl.js (hunk 3 above)
Comment (jallamsetty1) [reply modules/xmpp/SignalingLayerImpl.js]: Done

File: JitsiConference.js (hunk 1 above)
Comment (jallamsetty1) [reply JitsiConference.js:165]: A Map made more sense for the participant list, also it is going to make the transition to TS easier. I am trying to make small incremental changes whenever I am touching a class so that its easy to make the switch later.

Comment (hristoterezov) [reply JitsiConference.js:165]: I don't see how it could affect the transition to TS but I see how a map can be more clear as a type.

File: JitsiConference.js [hunk 4]
```diff
@@ -389,6 +395,17 @@ JitsiConference.prototype._init = function(options = {}) {
     );
 
     this._signalingLayer.setChatRoom(this.room);
+    this._signalingLayer.on(
+        SignalingEvents.SOURCE_UPDATED,
+        (sourceName, endpointId, muted, videoType) => {
```
Comment (saghul) [inline JitsiConference.js:400]: Why do we get the video type here, but for the media type we need to query it?

File: JitsiParticipant.js [hunk 5]
```diff
@@ -43,6 +43,7 @@ export default class JitsiParticipant {
         this._isReplacing = isReplacing;
         this._isReplaced = isReplaced;
         this._features = new Set();
+        this._sources = new Map();
```
Comment (saghul) [inline JitsiParticipant.js:46]: It would be good to have a comment here mentioning how the data is mapped.

File: JitsiParticipant.js [hunk 6]
```diff
@@ -61,6 +62,32 @@ export default class JitsiParticipant {
...
+
+        if (sourceByMediaType?.size) {
+            sourceByMediaType.set(sourceName, sourceInfo);
+
+            return;
+        }
+
+        sourceByMediaType = new Map();
+        sourceByMediaType.set(sourceName, sourceInfo);
+        this._sources.set(mediaType, sourceByMediaType);
```
Comment (saghul) [inline JitsiParticipant.js:88]: Am I reading this correctly, that we'll have this Map with media type (so audio, video) pointing to another map, with the source infos, indexes by source name?

Wouldn't this make it simpler?

```
this._sources = {
    audio: new Map(),
    video: new Map()
};
```

Since the keys for the outermost map are always constant, are they not?

File: JitsiParticipant.js [hunk 7]
```diff
@@ -130,6 +157,19 @@ export default class JitsiParticipant {
         return this._role;
     }
 
+    /**
+     * Returns the sources for a given media type.
+     * @param {string} mediaType The type of media, 'audio' or 'video'.
+     * @returns {Map<string, Object>}
+     */
+    getSources(mediaType) {
+        if (!mediaType) {
```
Comment (saghul) [inline JitsiParticipant.js:166]: When would be need this? It feels weird to have 2 types of returns...

File: JitsiParticipant.js [hunk 8]
```diff
@@ -130,6 +157,19 @@ export default class JitsiParticipant {
         return this._role;
     }
 
+    /**
+     * Returns the sources for a given media type.
+     * @param {string} mediaType The type of media, 'audio' or 'video'.
+     * @returns {Map<string, Object>}
```
Comment (saghul) [inline JitsiParticipant.js:163]: If the parameter is missing, the return type is different.

File: modules/xmpp/JingleSessionPC.js [hunk 9]
```diff
@@ -1793,115 +1794,88 @@ export default class JingleSessionPC extends JingleSession {
...
-                    }
+                    // Update the track with all the relevant info.
+                    track.setSourceName(source);
+                    track.setOwner(owner);
+                    if (mediaType === MediaType.VIDEO) {
+                        const type = videoType === 'CAMERA' ? VideoType.CAMERA : VideoType.DESKTOP;
 
-                    track.setSourceName(s.source);
-                    track.setOwner(s.owner);
+                        track._setVideoType(type);
```
Comment (saghul) [inline modules/xmpp/JingleSessionPC.js:1826]: why not assign videoType directly?

File: modules/xmpp/SignalingLayerImpl.js [hunk 10]
```diff
@@ -142,12 +142,14 @@ export default class SignalingLayerImpl extends SignalingLayer {
                 = this._remoteSourceState[endpointId] || (this._remoteSourceState[endpointId] = {});
 
             for (const sourceName of Object.keys(sourceInfoJSON)) {
+                let sourceChanged = false;
```
Comment (saghul) [inline modules/xmpp/SignalingLayerImpl.js:145]: General comment on this file. Can you pl make sure Spot screen sharing won't break with these changes?

**Review by saghul — COMMENTED** [review]
Left some comments, nothing major! PTAL.

File: JitsiConference.js (hunk 4 above)
Comment (jallamsetty1) [reply JitsiConference.js:400]: The media type can be inferred from the source-name using `getMediaTypeFromSourceName` but not the video type.

File: JitsiParticipant.js (hunk 7 above)
Comment (jallamsetty1) [reply JitsiParticipant.js:166]: JM calls both `getSources()` and `getSources('video')`.  I added the first case as an after thought. I will return the map here and the caller of the function can do further processing to get the video sources.

File: modules/xmpp/JingleSessionPC.js (hunk 9 above)
Comment (jallamsetty1) [reply modules/xmpp/JingleSessionPC.js:1826]: It is a matter of preference I guess. I try to avoid it if I need to update it outside of the class and there is a function available.

File: JitsiParticipant.js (hunk 6 above)
Comment (jallamsetty1) [reply JitsiParticipant.js:88]: Yes, it does make it simpler but then also creates the need for declaring a new type for this Object here (when we transition to TS) and in [JM](https://github.com/jitsi/jitsi-meet/blob/master/react/features/base/participants/types.ts#L54).  The current approach doesn't sound too complicated for me, we do it the same way for remote tracks in TPC but do let me know if you have a strong preference.

File: JitsiConference.js (hunk 4 above)
Comment (saghul) [reply JitsiConference.js:400]: 👍

File: JitsiParticipant.js (hunk 6 above)
Comment (saghul) [reply JitsiParticipant.js:100]: No strong preference, and since it's an internal thing it can be changed anytime if we have a change of heart.

**Review by saghul — APPROVED** [review]
